lualib/rspamadm/mime.lua

```diff
@@ -873,19 +879,92 @@ local function sign_handler(opts)
   end
 end
 
+-- Strips directories and .extensions (if present) from a filepath
+local function filename_only(filepath)
+  local filename = filepath:match(".*%/([^%.]+)")
+  if not filename then
+    filename = filepath:match("([^%.]+)")
+  end
+  return filename
+end
+
+assert(filename_only("very_simple") == "very_simple")
+assert(filename_only("/home/very_simple.eml") == "very_simple")
+assert(filename_only("very_simple.eml") == "very_simple")
+assert(filename_only("very_simple.example.eml") == "very_simple")
+assert(filename_only("/home/very_simple") == "very_simple")
+assert(filename_only("home/very_simple") == "very_simple")
+assert(filename_only("./home/very_simple") == "very_simple")
+assert(filename_only("../home/very_simple.eml") == "very_simple")
+assert(filename_only("/home/dir.with.dots/very_simple.eml") == "very_simple")
+
+--Write the dump content to file or standard out
+local function write_dump_content(dump_content, fname, extension, outdir)
+  local wrote_filepath = nil
+  if outdir then
+    if outdir:sub(-1) ~= "/" then
+      outdir = outdir .. "/"
+    end
+
+    local outpath = string.format("%s%s.%s", outdir, filename_only(fname), extension)
+    local outfile = io.open(outpath, "w")
+
+    if outfile then
+      outfile:write(dump_content)
+      outfile:close()
+      io.write(outpath.."\n")
+      wrote_filepath = outpath
+    else
+      io.stderr:write(string.format("Unable to open: %s\n", outpath))
+    end
+  else
+      io.write(dump_content)
+  end
+  return wrote_filepath
+end
+
+-- Get the formatted ucl (split or unsplit) or the raw task content
+local function get_dump_content(task, opts, fname)
```

Comment:
Understood. Presumably if `dump_content` within the `write_dump_content` is a string I would NOT load it into rspamd_text, and just write it to file to avoid the copy. If it is userdata I would load it into rspamd_text and use the save_in_file method? 

How would I instantiate rspamd_text from userdata. I thought `fromtable` but that appears to be incorrect. The docs appear unclear on this

Reply:
Methods to get content return userdata that is `rspamd_text`. In theory, you can ensure that `userdata` is a text by comparing it's `.cookie` field with `rspamd_text.cookie`. But if you operate with stuff returned from `task` methods, they can realistically return merely `rspamd_text` or a `string`. `rspamd_text` was designed for a sole purpose to provide opaque structure to prevent large strings interning in Lua which could be quite expensive and requires many copying when passing that between Lua and C.

Reply:
This approach of having `get_dump_content` not always returning a string (using tostring) complicates the writing a fair bit. The conditional type checks need to be duplicated in the "write to file" code and the "write to stdout" code.

I tried to normalise it by writing to a fd. Defaulting to 1, but set to the return value of `rspamd_util.create_file` if -o is specified. However, it appears that the rspamd code `save_in_file` and `create_file` do not support overwritting existing files in the same manner `io.open(.., "w")` does?

```
if type(dump_content) == "userdata" then
      if dump_content:save_in_file(outpath, "w") then
        wrote_filepath = outpath
      else
        io.stderr:write(string.format("Unable to save_in_file: %s\n", outpath))
      end
```

Getting `Unable to save_in_file: /tmp/Zipfile_test-part3.raw`  etc

Reply:
AFAIR, you can use `-` as a file name to write to stdout. But I agree that the best approach would be to accept io handle from Lua. Last time I have checked that it was quite difficult to implement though...

Reply:
I believe my latest iteration is a happy balance between performance and readability. In summary:

- Instead of `tostring` on userdata in `get_dump_content` the content is returned directly
- Within  `write_dump_content` string `dump_content` will be loaded into rspamd_text type

While there is an additional copy taken here during the `rspamd_text.fromstring` call I believe this is acceptable because:
- it is still going to be more performant than the old "userdata to string" approach
- It is inline with code elsewhere. For example: https://github.com/rspamd/rspamd/blob/a15e3d27f41d18d917f93c25a1f475f09ad14e41/lualib/lua_scanners/avast.lua#L125 
- It allows to rest of `write_dump_content` to not care about type, and use `save_in_file`